File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第10章出差.java
//出差 lanqiaoOJ 2194 

import java.io.*;
import java.util.*;


public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static Scanner cin = new Scanner(System.in);
    static int INF = 0x3f3f3f3f;
    static int M = 20010;
    static int[] t = new int[M];
    static int[] dist = new int[M];
    static int[] ea = new int[M];
    static int[] eb = new int[M];
    static int[] ec = new int[M];


    public static void main(String[] args) {
        int n = cin.nextInt(), m = cin.nextInt();
        for (int i = 1; i <= n; i++) t[i] = cin.nextInt();
        for (int i = 1; i <= m; i++) {
            int a = cin.nextInt(), b = cin.nextInt(), c = cin.nextInt();
            ea[i] = a;
            eb[i] = b;
            ec[i + m] = ec[i] = c;
            ea[i + m] = b;
            eb[i + m] = a;
        }
        Arrays.fill(dist, INF);
        dist[1] = 0;
        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= 2 * m; i++) {
                int u = ea[i], v = eb[i];
                int res = t[v];
                if (v == n) res = 0;
                dist[v] = Math.min(dist[v], dist[u] + ec[i] + res);
            }
        }
        System.out.println(dist[n]);
    }
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第10章出差.py
#出差 //lanqiaoOJ  2194

n, m = map(int, input().split())
t= [0]+[int(i) for i in input().split()]  #不用t[0]，从t[1]开始
e = []                 #简单的数组存图
for i in range(1,m+1):
    a, b,c = map(int, input().split())
    e.append([a,b,c]) 
    e.append([b,a,c])  #双向边
dist=[0x3f3f3f3f]*(n+1)
dist[1]=0
for k in range(1,n+1):
    for a,b,c in e:
        res=t[b]
        if b==n: res=0
        dist[b]=min(dist[b],dist[a]+c+res)
print(dist[n])
